ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/engine/CommandLine.kt

```diff
@@ -0,0 +1,158 @@
+
+        (argumentsMap["-watch"]?.split(",") ?: configuration.tryGetStringList(ConfigKeys.hostWatchPaths))?.let {
+            watchPaths = it
+        }
+    }
+
+    val host = argumentsMap["-host"] ?: configuration.tryGetString(ConfigKeys.hostConfigPath) ?: "0.0.0.0"
+    val port = argumentsMap["-port"] ?: configuration.tryGetString(ConfigKeys.hostPortPath)
+    val sslPort = argumentsMap["-sslPort"] ?: configuration.tryGetString(ConfigKeys.hostSslPortPath)
+    val sslKeyStorePath = argumentsMap["-sslKeyStore"] ?: configuration.tryGetString(ConfigKeys.hostSslKeyStore)
+    val sslKeyStorePassword = configuration.tryGetString(ConfigKeys.hostSslKeyStorePassword)?.trim()
+    val sslPrivateKeyPassword = configuration.tryGetString(ConfigKeys.hostSslPrivateKeyPassword)?.trim()
+    val sslKeyAlias = configuration.tryGetString(ConfigKeys.hostSslKeyAlias) ?: "mykey"
+
+    if (port == null && sslPort == null) {
+        throw IllegalArgumentException(
+            "Neither port nor sslPort specified. Use command line options -port/-sslPort " +
+                "or configure connectors in application.conf"
+        )
+    }
+
+    val engineConfig = BaseApplicationEngine.Configuration()
+    if (port != null) {
+        engineConfig.connector {
+            this.host = host
+            this.port = port.toInt()
+        }
+    }
+    if (sslPort != null) {
+        engineConfig.configureSSLConnectors(
+            host,
+            sslPort,
+            sslKeyStorePath,
+            sslKeyStorePassword,
+            sslPrivateKeyPassword,
+            sslKeyAlias
+        )
+    }
+
+    return CommandLineConfig(applicationProperties, engineConfig)
+}
+
+internal fun buildApplicationConfig(args: List<Pair<String, String>>): ApplicationConfig {
+    val commandLineProperties = args
+        .filter { it.first.startsWith("-P:") }
+        .map { it.first.removePrefix("-P:") to it.second }
+
+    val configPaths = args.filter { it.first == "-config" }.map { it.second }
+
+    val commandLineConfig = MapApplicationConfig(commandLineProperties)
+    val environmentConfig = getConfigFromEnvironment()
+
+    val fileConfig = when (configPaths.size) {
+        0 -> ConfigLoader.load()
+        1 -> ConfigLoader.load(configPaths.single())
+        else -> configPaths.map { ConfigLoader.load(it) }.reduce { first, second -> first.mergeWith(second) }
+    }
+
+    return fileConfig.mergeWith(environmentConfig).mergeWith(commandLineConfig)
+}
+
+internal expect fun ApplicationEngine.Configuration.configureSSLConnectors(
+    host: String,
+    sslPort: String,
+    sslKeyStorePath: String?,
+    sslKeyStorePassword: String?,
+    sslPrivateKeyPassword: String?,
+    sslKeyAlias: String
+)
+
+internal expect fun ApplicationEnvironmentBuilder.configurePlatformProperties(args: Array<String>)
+
+internal expect fun getConfigFromEnvironment(): ApplicationConfig
+
+/**
+ * Load engine's configuration suitable for all engines from [deploymentConfig]
```

Comment:
Small improvement: "Loads common engine configuration parameters applicable to all engine types from the specified [ApplicationConfig]."

And then add the param separately:
"@ param deploymentConfig The application configuration."